chain/pool/src/lib.rs

```diff
@@ -1,83 +1,121 @@
-    num_transactions: usize,
-    /// Transactions grouped by account and ordered by nonce.
-    pub transactions: HashMap<AccountId, BTreeMap<Nonce, SignedTransaction>>,
+    /// Transactions grouped by a pair of (account ID, signer public key).
+    /// It's more efficient to keep transactions unsorted and with potentially conflicting nonce
+    /// than create a BTreeMap for every transaction on average.
+    pub transactions: HashMap<(AccountId, PublicKey), Vec<SignedTransaction>>,
+    /// Set of all hashes to quickly check if the given transaction is in the pool.
+    pub unique_transactions: HashSet<CryptoHash>,
 }
 
 impl TransactionPool {
-    /// Insert a valid transaction into the pool that passed validation.
-    pub fn insert_transaction(&mut self, valid_transaction: ValidTransaction) {
-        let signer_id = valid_transaction.transaction.transaction.signer_id.clone();
-        let nonce = valid_transaction.transaction.transaction.nonce;
-        self.num_transactions += 1;
+    /// Insert a signed transaction into the pool that passed validation.
+    pub fn insert_transaction(&mut self, signed_transaction: SignedTransaction) {
+        if self.unique_transactions.contains(&signed_transaction.get_hash()) {
+            return;
+        }
+        self.unique_transactions.insert(signed_transaction.get_hash());
+        let signer_id = signed_transaction.transaction.signer_id.clone();
+        let signer_public_key = signed_transaction.transaction.public_key.clone();
         self.transactions
-            .entry(signer_id)
-            .or_insert_with(BTreeMap::new)
-            .insert(nonce, valid_transaction.transaction);
+            .entry((signer_id, signer_public_key))
+            .or_insert_with(Vec::new)
+            .push(signed_transaction);
     }
 
     /// Take transactions from the pool, in the appropriate order to be put in a new block.
-    /// Ensure that on average they will fit into expected weight.
+    /// We first take one transaction per key of (AccountId, PublicKey) with the lowest nonce,
+    /// then we take the next transaction per key with the lowest nonce.
     pub fn prepare_transactions(
         &mut self,
         expected_weight: u32,
     ) -> Result<Vec<SignedTransaction>, Error> {
-        // TODO: pack transactions better.
-        let result = self
-            .transactions
-            .values()
-            .flat_map(BTreeMap::values)
-            .take(expected_weight as usize)
-            .cloned()
-            .collect();
+        let mut sorted = false;
+        let mut result = vec![];
+        while result.len() < expected_weight as usize && !self.transactions.is_empty() {
+            let mut keys_to_remove = vec![];
+            for (key, txs) in self.transactions.iter_mut() {
+                if !sorted {
+                    // Reverse sort by nonce (non-increasing)
+                    txs.sort_by(|a, b| b.transaction.nonce.cmp(&a.transaction.nonce));
```

Comment:
Use specialization `sort_by_key`.

Reply:
`sort_by_key` sorts ascending order. Vec.pop_front is more expensive

Reply:
```rust
sort_by_key(|t| std::cmp::Reverse(t.nonce))
```